addons/account/models/account_journal.py

```diff
@@ -419,7 +419,7 @@ def write(self, vals):
                     bank_account = self.env['res.partner.bank'].browse(vals['bank_account_id'])
                     if bank_account.partner_id != company.partner_id:
                         raise UserError(_("The partners of the journal's company and the related bank account mismatch."))
-            if 'alias_name' in vals:
+            if 'alias_name' in vals  or vals.get('type') in ('purchase', 'sale') or (journal.type in ('purchase', 'sale') and not 'alias_id' in vals):
```

Comment:
Can you explain this part?
`or (journal.type in ('purchase', 'sale') and not 'alias_id' in vals)`